### PR TITLE
feat: add `synced` field

### LINKED ISSUE #25: Add synced/unsynced attribute

### PR DESCRIPTION
**Pull request by alessandromazza98**
Closes #25 

This PR adds the `synced` field, useful to distinguish between up to date with the latest blocks nodes and unsynced ones.

To do that we do the following:
- we use a `ethers` provider for RPC calls. (Note here that when ready, we can easily switch to `alloy`!)
- we do 2 RPC calls, one to retrieve the latest block number from the provider and one to retrieve (from the provider) the block number corresponding to the best block hash taken from the peer's status message (`status.blockhash`).
- then we compare the two block numbers like this:
```rust
if peer_best_block_number < last_block_number - SYNCED_THRESHOLD {
    synced = Some(false);
} else {
    synced = Some(true);
}
```
where `SYNCED_THRESHOLD` is just a const equals to 100 (blocks). Note that this could be changed if we want to use a different threshold block number.

`synced` is an `Option<bool>` because it can happen that those RPC calls do not succeed so we don't have this information. From local testing I see that the majority of peers have a `synced` flag set so I'm optimistic on this.

Also note that now I set the provider to use `"https://eth.llamarpc.com"` RPC endpoint which is a free one. Of course we **MUST** change it with a dedicated node that can always serve our RPC queries.

### CODE REVIEW

File: bins/reth-crawler/src/crawler/listener/update_listener.rs
Comment (0xprames) [inline bins/reth-crawler/src/crawler/listener/update_listener.rs:31]: looks ok to me, but cc @rjected to make sure

File: bins/reth-crawler/src/crawler/listener/update_listener.rs
Comment (0xprames) [inline bins/reth-crawler/src/crawler/listener/update_listener.rs:347]: if this call fails for w/e reason, we just skip setting this flag altogether is how im reading this

that should be fine, but how do we handle storing a missing flag in SQLite?

File: db/src/db.rs
Comment (0xprames) [inline db/src/db.rs:91]: i think this line answers my above question

File: bins/reth-crawler/src/crawler/listener/update_listener.rs
Comment (alessandromazza98) [reply bins/reth-crawler/src/crawler/listener/update_listener.rs:347]: we handle it right now.

If you see the Sqlite db I did not put `NOT NULL`. So if we do not have this info, sqlite simply displays `NULL`.

**Comment by 0xprames** [comment]
looks fine to me - as discussed we probably want to:

- this shows synced at the last_seen time, and does not guarantee that the attribute is kept fresh (specify in README and maybe on the website somewhere)
- need to replace the LllamaRPC endpoint with the reth team node endpoint
- this will require a backfill for all existing data populated before we roll this out

let's wait on @rjected to review this as well before merging

**Comment by 0xprames** [comment]
can we add local test details to show this works? @alessandromazza98

**Comment by alessandromazza98** [comment]
Test on my mac:

Command:
Screenshot 2023-11-06 at 17 42 06

SQLite db (after some minutes), note the `synced` attribute:
Screenshot 2023-11-06 at 17 43 22

Note that in SQLite when the `synced` attribute is `None` it is displayed as `NULL`

File: bins/reth-crawler/src/crawler/listener/update_listener.rs
Comment (Rjected) [inline bins/reth-crawler/src/crawler/listener/update_listener.rs:202]: id make an issue to get rid of the duplication between the code here and the code in `start_dnsdisc`

**Review by Rjected — APPROVED** [review]
Approving so not block this, but would like some of the comments addressed

File: db/src/types.rs
Comment (alessandromazza98) [reply db/src/types.rs:27]: if it's only a `bool` how should we set the field on the database? false by default?

**Comment by Rjected** [comment]
I guess the synced field is an Option to be backwards compatible, and backfill seems like it would need to happen whenever you're migrating and have new fields

File: bins/reth-crawler/src/crawler/listener/update_listener.rs
Comment (alessandromazza98) [reply bins/reth-crawler/src/crawler/listener/update_listener.rs:202]: yes. On this point I just copy paste here because we can later refactor all the `update_listener.rs` file since a there is a lot of duplication in that file.

File: db/src/types.rs
Comment (Rjected) [reply db/src/types.rs:27]: > if it's only a `bool` how should we set the field on the database? false by default?

Yes, or at least it was my understanding that you can't insert into the database until `PeerData` is constructed, and it is only ever constructed after we determine whether or not the peer is synced (when `synced` is set to `Some`). I guess `None` is valid if there is no working RPC though, since there would be no way to know if the peer is synced.

Comment (alessandromazza98) [reply db/src/types.rs:27]: exactly. So I used `Option<bool>` just in order to be able to represent the fact that we don't know if that node is synced or not since we weren't able to get this information, so we just save `None` which gets translated into a `NULL`

**Pull request merged**
